src/goto-instrument/synthesizer/cegis_verifier.cpp

```diff
@@ -0,0 +1,482 @@
+  // Loop-entry valuations of havoced variables.
+  std::map<exprt, std::size_t> loop_entry_values;
+  std::map<exprt, std::size_t> loop_entry_offsets;
+
+  // Live variables upon the loop head.
+  std::set<exprt> live_variables;
+
+  bool entered_loop = false;
+
+  // Scan the trace step by step to store needed valuations.
+  for(const auto &step : goto_trace.steps)
+  {
+    switch(step.type)
+    {
+    case goto_trace_stept::typet::DECL:
+    case goto_trace_stept::typet::ASSIGNMENT:
+    {
+      if(!step.full_lhs_value.is_nil())
+      {
+        // Entered loop?
+        if(is_assignment_to_instrumented_variable(step.pc, ENTERED_LOOP))
+          entered_loop = step.full_lhs_value == true_exprt();
+
+        // skip hidden steps
+        if(step.hidden)
+          break;
+
+        // Live variables
+        // 1. must be in the same function as the target loop;
+        // 2. alive before entering the target loop;
+        // 3. a pointer or a primitive-typed variable;
+        // TODO: add support for union pointer
+        if(
+          step.pc->source_location().get_function() ==
+            loop_entry_loc.get_function() &&
+          !entered_loop &&
+          (step.full_lhs.type().id() == ID_unsignedbv ||
+           step.full_lhs.type().id() == ID_pointer) &&
+          step.full_lhs.id() == ID_symbol)
+        {
+          const auto &symbol =
+            expr_try_dynamic_cast<symbol_exprt>(step.full_lhs);
+
+          // malloc_size is not-hidden tmp variable.
+          if(id2string(symbol->get_identifier()) != "malloc::malloc_size")
+          {
+            live_variables.emplace(step.full_lhs);
+          }
+        }
+
+        // Record valuation of primitive-typed variable.
+        if(step.full_lhs.type().id() == ID_unsignedbv)
+        {
+          // Store the value into the map for loop_entry value if we haven't
+          // entered the loop.
+          if(!entered_loop)
+          {
+            loop_entry_values[step.full_lhs] =
+              hex_to_size_t(step.full_lhs_value.get_string(ID_value));
+          }
+
+          // Store the value into the the map for havoced value if this step
+          // is a loop havocing instruction.
+          if(is_loop_havoc(step.pc))
+          {
+            havoced_values[step.full_lhs] =
+              hex_to_size_t(step.full_lhs_value.get_string(ID_value));
+          }
+        }
+
+        // Record object_size, pointer_offset, and loop_entry(pointer_offset).
+        if(
+          can_cast_expr<annotated_pointer_constant_exprt>(
+            step.full_lhs_value) &&
+          contains_symbol_prefix(
+            step.full_lhs_value, SYMEX_DYNAMIC_PREFIX "dynamic_object"))
+        {
+          const auto &annotated_pointer_constant_expr =
+            expr_try_dynamic_cast<annotated_pointer_constant_exprt>(
+              step.full_lhs_value);
+
+          object_sizes[step.full_lhs] =
+            extract_size(annotated_pointer_constant_expr->symbolic_pointer());
```

Comment:
Use `pointer_arithmetict` to pick up the pointer, and do `pointer_offset_size(to_pointer_type(pointer.type().base_type(), ns))`; you'll then want to use `numeric_cast<std::size_t>` of the result.